### PR TITLE
KVP-MAX-READ: Update to run on Azure + HyperV

### PR DESCRIPTION
**Pull request by stefanlupsa**
Dropping the PR as requirement for the test is that the KVP is read from the host so this cannot be ported to run on Azure.
Without reading the kvp pools from the host side, the original bug check does not repro, making the test not relevant in Azure to run.

---

* use kvp client tool to for read
* port functionality to bash script to be compatible with Azure and HyperV
* use kvp pool 4 as it's empty on Azure

### CODE REVIEW

**Comment by stefanlupsa** [comment]
HyperV: 
```
[LISAv2 Test Results Summary]
Test Run On           : 05/29/2019 17:17:54
VHD Under Test        : Ubuntu-18.04-gen1.vhdx
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:1

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 KVP                  KVP-MAX-READ                                                                      PASS                 0.42
```
Azure:
```
[LISAv2 Test Results Summary]
Test Run On           : 05/29/2019 17:19:51
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:3

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 KVP                  KVP-MAX-READ                                                                      PASS                 0.76
```

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+# kvp_client can output max 200 entries
+expectedEentryCount="$Entries"
+if [ "$expectedEentryCount" -gt 200 ]; then
+    expectedEentryCount=200
+fi
+kvp_client_output="$(./${kvp_client} ${Pool})"
+
+if [ "$(echo "${kvp_client_output}" | grep -c "Pool is ${Pool}")" -ne 1 ]; then
+    LogErr "Wrong pool otuput"
```
Comment (LiliDeng) [inline Testscripts/Linux/KVP-MaxRead.sh]: otuput => output

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+done
+
+if [ $? -ne 0 ]; then
+    LogErr "Failed to replace new entries"
+    SetTestStateFailed
+    exit 0
+fi
+
+# kvp_client can output max 200 entries
+expectedEentryCount="$Entries"
```
Comment (LiliDeng) [inline Testscripts/Linux/KVP-MaxRead.sh]: expectedEentryCount => expectedEntryCount

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+if [ $(echo "${kvp_client_output}" | grep -c "Key:.*Value:.*") -ne "$expectedEentryCount" ]; then
+    LogErr "Number of Entries doesn't match $expectedEentryCount"
+    LogErr "${kvp_client_output}"
+    SetTestStateFailed
+    exit 0
+fi
+
+if [ $(ps aux | grep -c [k]vp) -lt 1 ]; then
+    LogErr "KVP daemon failed after append"
```
Comment (LiliDeng) [inline Testscripts/Linux/KVP-MaxRead.sh]: Could you please provide meaningful message?

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+# Append Stage 2
+# kvp_client also deletes entries when appending for the same key
+counter=0
+while [ "$counter" -le "$Entries" ]; do
+    ./"${kvp_client}" append "$Pool" "${key}${counter}" "${value}"
+    let counter=counter+1
+done
+
+if [ $? -ne 0 ]; then
```
Comment (LiliDeng) [inline Testscripts/Linux/KVP-MaxRead.sh]: here $?, what is standing for?

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 2 above)
Comment (LiliDeng) [reply Testscripts/Linux/KVP-MaxRead.sh]: if the max is 200, why define it in xml as 250?

**Review by LiliDeng — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+# Licensed under the Apache License.
+
+# Source utils.sh
+. utils.sh || {
+    echo "ERROR: unable to source utils.sh!"
+    echo "TestAborted" > state.txt
+    exit 0
+}
+UtilsInit
+
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh:12]: double line?

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+if [ "$(ps aux | grep -c [k]vp)" -lt 1 ]; then
+    LogErr "KVP daemon is not running"
+    SetTestStateAborted
+    exit 0
+fi
+
+# Verify OS architecture, select kvp client
+if [ "$(uname -a | grep -c x86_64)" -eq 1 ]; then
+    LogMsg "64 bit architecture was detected"
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh]: 64-bit

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    kvp_client="kvp_client64"
+elif [ "$(uname -a | grep -c i686)" -eq 1 ]; then
+    LogMsg "32 bit architecture was detected"
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh]: 32-bit

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    kvp_client="kvp_client32"
+else
+    LogErr "Unable to detect OS architecture"
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh]: "Unable to detect OS architecture or unsupported architecture" It is nice to display 'uname -a' in the log for debugging purpose.

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+# Append stage 1
+value="value"
+key="test"
+counter=0
+while [ "$counter" -le "$Entries" ]; do
+    ./"${kvp_client}" append "$Pool" "${key}${counter}" "${value}"
+    let counter=counter+1
+done
+
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh:45]: double line?

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+if [ $? -ne 0 ]; then
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh]: previous operation could be 'done' or 'let' statement. re-design the check-out.

File: XML/TestCases/FunctionalTests-KVP.xml [hunk 11]
```diff
@@ -118,15 +118,15 @@
     </test>
     <test>
         <testName>KVP-MAX-READ</testName>
-        <testScript>KVP-MaxRead.ps1</testScript>
-        <files>.\Tools\kvp_client32,.\Tools\kvp_client64</files>
+        <testScript>KVP-MaxRead.sh</testScript>
+        <files>.\Testscripts\Linux\KVP-MaxRead.sh,.\Testscripts\Linux\utils.sh,.\Tools\kvp_client64,.\Tools\kvp_client32</files>
```
Comment (juhlee-microsoft) [inline XML/TestCases/FunctionalTests-KVP.xml:122]: please make sure if those new binaries are available in blob and network share. Confirm with @adriansuhov

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+. utils.sh || {
+    echo "ERROR: unable to source utils.sh!"
+    echo "TestAborted" > state.txt
+    exit 0
+}
+UtilsInit
+
+
+if [ "$(ps aux | grep -c [k]vp)" -lt 1 ]; then
+    LogErr "KVP daemon is not running"
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh:14]: if this statement is validating KVP daemon is not running, then outer- if-statement meant 'KVP daemon exists'. It is nice to log those information for debugging.

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    LogErr "Unable to detect OS architecture"
+    SetTestStateAborted
+    exit 0
+fi
+
+# Append stage 1
+value="value"
+key="test"
+counter=0
+while [ "$counter" -le "$Entries" ]; do
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh:36]: Add log msg about what it is trying to do.

File: Testscripts/Linux/KVP-MaxRead.sh [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+if [ "$(echo "${kvp_client_output}" | grep -c "Pool is ${Pool}")" -ne 1 ]; then
+    LogErr "Wrong pool otuput"
+    LogErr "${kvp_client_output}"
+    SetTestStateFailed
+    exit 0
+fi
+
+# expect more entries than client can output
+if [ "$Entries" -gt "$expectedEentryCount" ] && \
+    [ "$(echo "$kvp_client_output" | grep -c "Num records is 200")" -lt 1 ] && \
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/KVP-MaxRead.sh:75]: that '200' meant expectedEentryCount? no need to keep constant in the log. will be harder maintenance.

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 5 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh:12]: done

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 6 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: done

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 7 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: done

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 10 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: done

File: XML/TestCases/FunctionalTests-KVP.xml (hunk 11 above)
Comment (stefanlupsa) [reply XML/TestCases/FunctionalTests-KVP.xml:122]: they are, the old test used them too

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 4 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: fixed by moving the check in the loop block

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 1 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: done

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 3 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: Not really, this was part of the old test, after adding the new KVP entries it checked that the daemon was still running, and is a condition for failing.

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 8 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: done

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 9 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh:45]: done

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 3 above)
Comment (LiliDeng) [reply Testscripts/Linux/KVP-MaxRead.sh]: KVP daemon exit abnormally?

Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: Ah, rephrase, I thought you meant add more content.

Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: done

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 2 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh]: The max that can be read by kvp_client is 200, more entries can be added but you can't read more

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 12 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh:14]: I don't think this is relevant. this is part of prerequisite checking, in this case we need the daemon to be running. The daemon binaries could be present but not from the current kernel for example, it's too much cost to check something like this here and for little value.

File: Testscripts/Linux/KVP-MaxRead.sh (hunk 13 above)
Comment (stefanlupsa) [reply Testscripts/Linux/KVP-MaxRead.sh:36]: added log message after each append

**Comment by stefanlupsa** [comment]
Dropping the PR as requirement for the test is that the KVP is read from the host so this cannot be ported to run on Azure as this functionality is not supported

**Pull request closed without merging**